### ISSUE TITLE
Broken Link Checker

### ISSUE THREAD

**Issue by AngelaFaulding** [NHSDigital/DataDictionaryPublication#310]
A broken link checker used to be run on the NHS Data Model and Dictionary website. Any broken links could then be updated.
The broken link checker no longer runs so there is currently no way of identifying broken links until they are spotted by a member of the team or via the helpdesk.
There may be a large number of broken links this year following the NHS Digital / NHS X / NHS England merge.
James said he may be able to incorporate a broken link checker into Mauro.

**Comment by AngelaFaulding** [comment]
To be looked at Post Migration.

**Comment by AngelaFaulding** [comment]
James has been able to do this but it wouldn't work in the meeting on 18/1/22.
Next meeting planned for 25/1/22.

**Comment by GSChandel** [comment]
James - It's working.
DD Team to test and confirm.

**Comment by AngelaFaulding** [comment]
This will only be resolved when #392 is fixed.

**Comment by AngelaFaulding** [comment]
The latest release now has no failures in the Orchestrator which relates to #451

**Comment by AngelaFaulding** [comment]
The Broken Link Checker has been removed from the latest version of the Orchestrator. Can this please be reinstated as it was originally agreed and working.

July 2024:

![image](https://github.com/NHSDigital/DataDictionaryPublication/assets/62995711/e75ba47a-a6e6-4c26-a32c-6b9f3ef7388d)

Previous version sorted by James:

![image](https://github.com/NHSDigital/DataDictionaryPublication/assets/62995711/11a81b2e-7c33-4bf2-b0f6-e64ed058a16c)

**Comment by pjmonks** [comment]
For clarification, System C are not responsible for a broken link checker integrity check. This is not mentioned in our statement of work to complete, so we consider this out of scope.

However, this may be partially related (unless the broken link checker refers to something else). We do have a criteria to uphold here:

> ED10.1   Ability created to rename a data item and have that change be reflected in any links recorded in other data item descriptions.

By renaming an item in a branch, there could then be the potential for hyperlinks in other description fields to become broken since the name in item paths have changed. This is in progress, currently in a testing state, as described in this pull request: MauroDataMapper-NHSD/mdm-plugin-nhs-data-dictionary#50

For this use case, System C are working on maintaining links to other items so there are no broken links. Anything beyond this is out of scope.

**Comment by AngelaFaulding** [comment]
16/07/2024 - James has tested this as part of the Orchestrator and has sent it to System C to check. Peter will deploy if it works on the server.

**Comment by AngelaFaulding** [comment]
When clicking on an item identified as having a broken link, the following is displayed. This appears to be just for Terms, i.e. the NHS Business Definitions. The links to the Data Elements and Attributes work.

![image](https://github.com/user-attachments/assets/9dd7e66e-d7bb-4311-894f-1ca852681535)

The text is:
## Not Found

We're sorry, but the server returned a 'Not Found' error

## Details

```json
{
  "headers": {
    "normalizedNames": {},
    "lazyUpdate": null
  },
  "status": 404,
  "statusText": "OK",
  "url": "https://mauro.dev.dataproducts.nhs.uk/api/terminologies/null",
  "ok": false,
  "name": "HttpErrorResponse",
  "message": "Http failure response for https://mauro.dev.dataproducts.nhs.uk/api/terminologies/null: 404 OK",
  "error": {
    "path": "/api/terminologies/null",
    "resource": "Terminology",
    "id": "null"
  }
}
```

Also, I have check ed a few of the items and they do not contain a broken link, for example: https://www.datadictionary.nhs.uk/nhs_business_definitions/british_hiv_association.html

**Comment by pjmonks** [comment]
Please could you identify which item you were intending to view in Mauro? The error message is not descriptive enough to pinpoint how the error happened.

I cannot speak to the broken link checker as that is the responsibility of @jamesrwelch , however it might be possible that links to external sites may work but only because they are redirected, you may be using an old link that just happens to work because the target website automatically redirects you to the updated page. I suppose it is up to you whether you consider those links "broken" or at least maintained.

**Comment by AngelaFaulding** [comment]
I just clicked from the top down and all NHS Business Definitions failed.

**Comment by jamesrwelch** [comment]
Also worth noting that links can be broken temporarily - it checks for any 4** error response code which can include time-outs, rate-limits, etc.

**Comment by AngelaFaulding** [comment]
Thanks @jamesrwelch - I have seen that before. So we would check the link and if it is OK, we wouldn't have to do anything with it.

The NHS Business Definitions not displaying is more of an issue.

**Comment by jamesrwelch** [comment]
Yes, given the age of some of the websites that are being pointed at, I wouldn't be surprised if you got a few that fail every once in a while.  If you get one that persistently fails in the integrity checker but works when you load it manually (or vice-versa), then we can take a closer look - for speed reasons the way we check for broken links is slightly different to loading it in a browser.
The links to business definitions being incorrect should be an easy fix that I'll look into.

**Comment by pjmonks** [comment]
> Please could you identify which item you were intending to view in Mauro? The error message is not descriptive enough to pinpoint how the error happened.

I mean which item from integrity check list were you clicking to visit in Mauro. I'm assuming you are:

1. Running the integrity checks.
2. They complete and show the "Broken links" section with a list of dictionary items that supposedly contain broken links.
3. You can clicked on of them to view and cannot find it in Mauro. Which of those items did you click on?

**Comment by AngelaFaulding** [comment]
I ran the integrity check for CR1815 and this is the Orchestrator page:

![image](https://github.com/user-attachments/assets/9948d9c5-8e05-4a07-85f8-34c4286e8f9e)

I clicked on each NHS Business Definition and got:

![image](https://github.com/user-attachments/assets/108d5a8d-8b20-45ca-b243-b5e04408afda)

**Comment by pjmonks** [comment]
I found the issue is not the broken link checker itself, the problem is the hyperlinks in the Orchestrator UI, which will be simple to fix. I've created this issue for tracking progress:

- MauroDataMapper-NHSD/nhsd-datadictionary-orchestrator#26

**Comment by pjmonks** [comment]
The hyperlinks in the integrity checker have been fixed now and deployed to test.
